### ISSUE TITLE
404 Not Found

### ISSUE THREAD

**Issue by barqers** [firefly-iii/firefly-iii#4156]
I am running Firefly III version 5.4.6

**Description**
Upgrading from 5.3.3 to 5.4.6 with the same docker compose broke the web page, receives only 404 Not Found. I've tried changing the workdir from /var/www/firefly-iii to /var/www/html and still get the same issue.

**Extra info**
Issue in Chrome, Firefox, Safari as well.

**Bonus points**

```
Now in entrypoint.sh for Firefly III


Entrypoint script version is 1.0.10 (2020-09-23)


Running as www-data.


Current working dir is /var/www/html


Now parsing _FILE variables.


done!


Touch DB file (if SQLite)...


Dump auto load...


Discover packages...


Wait for the database.


wait-for-it.sh: waiting 60 seconds for 192.168.x.x:3306


wait-for-it.sh: 192.168.x.x:3306 is available after 0 seconds


DB is up. Time to execute artisan commands.


Run various artisan commands...


Running migration commands...


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Running upgrade commands...


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Running verification commands...


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


nning report commands...


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Generating new OAuth keys...


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Go!


AH00112: Warning: DocumentRoot [/var/www/html/public] does not exist


AH00558: apache2: Could not reliably determine the server's fully qualified domain name, using 172.17.0.8. Set the 'ServerName' directive globally to suppress this message


AH00112: Warning: DocumentRoot [/var/www/html/public] does not exist


AH00558: apache2: Could not reliably determine the server's fully qualified domain name, using 172.17.0.8. Set the 'ServerName' directive globally to suppress this message


[Tue Dec 15 01:39:49.370301 2020] [mpm_prefork:notice] [pid 1] AH00163: Apache/2.4.38 (Debian) configured -- resuming normal operations


[Tue Dec 15 01:39:49.371045 2020] [core:notice] [pid 1] AH00094: Command line: 'apache2 -D FOREGROUND'


192.168.x.x - - [15/Dec/2020:01:39:49 +0000] "GET / HTTP/1.1" 404 397 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.198 Safari/537.36"


192.168.x.x - - [15/Dec/2020:01:39:50 +0000] "GET / HTTP/1.1" 404 396 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.198 Safari/537.36"


192.168.x.x - - [15/Dec/2020:01:39:51 +0000] "GET / HTTP/1.1" 404 396 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.198 Safari/537.36"
```

**Comment by JC5** [comment]
Delete the old image completely. Then retry.

**Comment by barqers** [comment]
> Delete the old image completely. Then retry.

Still seem to have the exact same issues as above, this is my current environmental variables below:
![Docker_Environmental_Variables](https://user-images.githubusercontent.com/34695760/102243018-b2424180-3ec8-11eb-9de1-2c7b7e3fb41a.png)

**Comment by JC5** [comment]
Unraid, right? Drop the firefly iii path. Drop app url. Otherwise I can't help you I'm afraid.

**Comment by JC5** [comment]
Also drop path.

**Comment by barqers** [comment]
> Unraid, right? Drop the firefly iii path. Drop app url. Otherwise I can't help you I'm afraid.

Yeah unraid, setting this up via Portainer. I remove app url, path, and firefly_iii_path, however those seem to get added on their own when I build the container.

I'm having a hard time reading the log to figure out if it's a DB issue or a file drive issue? Do the artisan commands run in the db?

**Comment by JC5** [comment]
File drive something. The script is here: https://github.com/firefly-iii/docker/blob/main/entrypoint.sh#L114

As far as I can tell it switches / changes the PWD out of the blue from that line on. 

I could use your help. Switch to the `:develop` image (see [this page](https://travis-ci.com/github/firefly-iii/docker/builds/208891377)) and see if the logs are more clear.

I changed nothing except some debug statements.

**Comment by barqers** [comment]
> File drive something. The script is here: https://github.com/firefly-iii/docker/blob/main/entrypoint.sh#L114
> 
> As far as I can tell it switches / changes the PWD out of the blue from that line on.
> 
> I could use your help. Switch to the `:develop` image (see [this page](https://travis-ci.com/github/firefly-iii/docker/builds/208891377)) and see if the logs are more clear.
> 
> I changed nothing except some debug statements.

Hmmm, seems to be largely the same problem and error messages:
`Now in entrypoint.sh for Firefly III


Entrypoint script version is 1.0.10 (2020-09-23)


Running as www-data.


Current working dir is /var/www/html


Now parsing _FILE variables.


done!


Touch DB file (if SQLite)...


Dump auto load...


Discover packages...


Wait for the database.


wait-for-it.sh: waiting 60 seconds for 192.168.x:3306


wait-for-it.sh: 192.168.x:3306 is available after 1 seconds


DB is up. Time to execute artisan commands.


Run various artisan commands...


Running migration commands...


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Running upgrade commands...


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Running verification commands...


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


nning report commands...


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Generating new OAuth keys...


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Go!


AH00112: Warning: DocumentRoot [/var/www/html/public] does not exist


AH00558: apache2: Could not reliably determine the server's fully qualified domain name, using 172.17.x. Set the 'ServerName' directive globally to suppress this message


AH00112: Warning: DocumentRoot [/var/www/html/public] does not exist


AH00558: apache2: Could not reliably determine the server's fully qualified domain name, using 172.17.x. Set the 'ServerName' directive globally to suppress this message


[Tue Dec 15 18:20:26.749515 2020] [mpm_prefork:notice] [pid 1] AH00163: Apache/2.4.38 (Debian) configured -- resuming normal operations


[Tue Dec 15 18:20:26.749694 2020] [core:notice] [pid 1] AH00094: Command line: 'apache2 -D FOREGROUND'


192.168.1.4 - - [15/Dec/2020:18:20:36 +0000] "GET /login HTTP/1.1" 404 397 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.198 Safari/537.36"


192.168.1.4 - - [15/Dec/2020:18:20:37 +0000] "GET /login HTTP/1.1" 404 397 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.198 Safari/537.36"`

Let me see if I can change the ownership of the files/folders, but I even tried running the container in privilege mode and it spits the same error messages out.

**Comment by JC5** [comment]
Yeah that's still the old script. Try again pls.

**Comment by barqers** [comment]
> Yeah that's still the old script. Try again pls.

Okay so I switched back to the 5.3.3 image and it boots back up no problem, login and I can see the data.

Then I repoint to :develop, and I get the errors above again. My entrypoint is "/usr/local/bin/entrypoint.sh":

`Run various artisan commands...


Running migration commands...


Could not open input file: artisan


Could not open input file: artisan


Could not open input file: artisan


Running upgrade commands...`

**Comment by barqers** [comment]
> Let me try again.

Yeah it's still mentioning the entrypoint script is 1.0.10 not 1.0.11 - did the updated entrypoint.sh script make it into the developer build? Not sure how to double check that.

**Comment by JC5** [comment]
The build failed but exited with 0. Stay tuned, sorry.

**Comment by JC5** [comment]
Alright the build succeeded, and the `:develop` image should include extra references to its working dir.

**Comment by barqers** [comment]
> Alright the build succeeded, and the `:develop` image should include extra references to its working dir.

Whoa, okay we're definitely getting somewhere, the log is miles long now. Shows entrypoint as 1.0.11, and looks like there's an issue with /var/www/html/bootstrap/cache access/ownership. If I run the container as user 'www-data', the container fails, if I run it as root with privileges then it actually works again.

So maybe the /var/www/html folder just needs to be chown'd to root:www-data instead of root:root?

`Now in entrypoint.sh for Firefly III,
Entrypoint script version is 1.0.11 (2020-12-15),
Running as www-data.,
Current working dir is /var/www/html,
Now parsing _FILE variables.,
done!,
Touch DB file (if SQLite)...,
Dump auto load...,
Discover packages...,
Current working dir is '/var/www/html',
Wait for the database.,
wait-for-it.sh: waiting 60 seconds for 192.168.x:3306,
wait-for-it.sh: 192.168.x:3306 is available after 0 seconds,
DB is up. Time to execute artisan commands.,
Current working dir is '/var/www/html',
Run various artisan commands...,
Running migration commands...,
[2020-12-16 14:48:18] local.ERROR: The /var/www/html/bootstrap/cache directory must be present and writable. {"exception":"[object] (Exception(code: 0): The /var/www/html/bootstrap/cache directory must be present and writable. at /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/PackageManifest.php:177),
[stacktrace],
#0 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/PackageManifest.php(136): Illuminate\\Foundation\\PackageManifest->write(),
#1 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/PackageManifest.php(106): Illuminate\\Foundation\\PackageManifest->build(),
#2 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/PackageManifest.php(89): Illuminate\\Foundation\\PackageManifest->getManifest(),
#3 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/PackageManifest.php(78): Illuminate\\Foundation\\PackageManifest->config(),
#4 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Bootstrap/RegisterFacades.php(26): Illuminate\\Foundation\\PackageManifest->aliases(),
#5 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Application.php(230): Illuminate\\Foundation\\Bootstrap\\RegisterFacades->bootstrap(),
#6 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Console/Kernel.php(310): Illuminate\\Foundation\\Application->bootstrapWith(),
#7 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Console/Kernel.php(127): Illuminate\\Foundation\\Console\\Kernel->bootstrap(),
#8 /var/www/html/artisan(37): Illuminate\\Foundation\\Console\\Kernel->handle(),,
},
"} ,
,
In PackageManifest.php line 177:,
                                                                             ,
  The /var/www/html/bootstrap/cache directory must be present and writable.  ,
                                                                             ,
,`

**Comment by JC5** [comment]
In your case, maybe, but I don't know how unraid handles container's access rights. Still on my list to improve this but I'm afraid you're on your own from this point on.

**Comment by barqers** [comment]
> In your case, maybe, but I don't know how unraid handles container's access rights. Still on my list to improve this but I'm afraid you're on your own from this point on.

Sounds good - going to run the container as root moving forward. When looking at 5.3.3 the directory /var/www/ was permitted for www-data so it looks like a recent change in the docker itself. Either way this is a working solution! Thanks again. Dev build looks good, no issues importing so far.

**Comment by Chaoschaot232** [comment]
Have the same problem but can not rly follow what was done to resolve it?
It is a clean and fresh install with docker + traefik and all what I am get is 404 error like stated above.

**Comment by barqers** [comment]
> Have the same problem but can not rly follow what was done to resolve it?
> It is a clean and fresh install with docker + traefik and all what I am get is 404 error like stated above.

Are you installing via docker-cli or a graphical tool like Portainer? Essentially the user inside the container has to be root because of the permissions set on /var/www **inside** the docker container. You can bash into a docker container using instructions like [https://www.edureka.co/community/10588/how-can-i-run-a-docker-exec-command-inside-a-docker-container](url)

Once inside if you execute some of the following commands:
`cd /var
ls -l
cd /www
ls -l
exit`

In my case, both outputs of ls -l returned root:root as the user of the folders and subfolders. Once you're back in your linux machine and out of the Docker container itself, you can check what user ran the Docker container:
`docker inspect $(docker ps -aq) --format '{{.Config.User}} {{.Name}}'
`

This should tell you whether you've properly ran it as the right user that matches the output of ls -l. If it's not the correct user then you can recreate the docker container but using the -u flag and the proper uid to match the user of /var/www (in my case it was root).

**Comment by Chaoschaot232** [comment]
> Are you installing via docker-cli or a graphical tool like Portainer?

I let all my containers startup over my own gitlab with gitlab runners.
I'll checkout the outputs of your cmd's and will report back soon :)
